File: ex00/ex00.cpp
#include "../incs/Matrix.hpp"

int main(void) {
	ft::Vector<float> v1({2., 3.});
	ft::Vector<float> v2({5., 7.});
	v1.print("v1");
	v2.print("v2");
	v1.add(v2);
	v1.print("add");
	// [7.]
	// [10.]

	v1 = ft::Vector<float>({2., 3.});
	v1.sub(v2);
	v1.print("sub");
	// [-3.]
	// [-4.]

	v1 = ft::Vector<float>({2., 3.});
	v1.scl(2.);
	v1.print("scl (*2)");
	// [4.]
	// [6.]

	ft::Matrix<float> m1({{1., 2.}, {3., 4.}});
	ft::Matrix<float> m2({{7., 4.}, {-2., 2.}});
	m1.print("m1");
	m2.print("m2");
	m1.add(m2);
	m1.print("add");
	// [8., 6.]
	// [1., 6.]

	m1 = ft::Matrix<float>({{1., 2.}, {3., 4.}});
	m1.sub(m2);
	m1.print("sub");
	// [-6., -2.]
	// [5., 2.]

	m1 = ft::Matrix<float>({{1.0, 2.0}, {3.0, 4.0}});
	m1.scl(2.0);
	m1.print("scl (*2)");
	// [2.0, 4.0]
	// [6.0, 8.0]

	return (0);
}

File: ex00/ex00.rs
struct Vector<'a, K> {
	arr: &'a mut [K],
}

impl<'a, K: std::ops::AddAssign + std::fmt::Display> Vector<'a, K> {
	fn from(init_array: &'a [K]) -> Self {
		Self {
			arr: init_array,
		}
	}
	fn add(&mut self, v: &Vector<K>) -> &mut Self {
		if self.arr.len() != v.arr.len() {
			return self;
		}
		for i in 0..self.arr.len() {
			self.arr[i] += v.arr[i];
		}
		return self;
	}

	// fn sub(&mut self, v: &Vector<K>) {

	// }

	// fn scl(&mut self, a: K) {
		
	// }
	pub fn Display(& self) {
		for i in 0..self.arr.len() {
			println!("{}", self.arr[i]);
		}
	}
}

fn main() {
	let mut u = Vector::from(&[2., 3.]);
	let v = Vector::from(&[5., 7.]);
	u.add(&v);
	u.Display();
	// [7.0]
	// [10.0]

	// let mut u = Vector::from([2., 3.]);
	// let v = Vector::from([5., 7.]);
	// u.sub(v);
	// println!("{}", u);
	// // [-3.0]
	// // [-4.0]

	// let mut u = Vector::from([2., 3.]);
	// u.scl(2.);
	// println!("{}", u);
	// // [4.0]
}